src/SourceLocator/Type/ClosureSourceLocator.php

```diff
@@ -4,100 +4,159 @@
-            $closureData['reflection']->getFileName()
-        );
+        $fileName = $this->coreFunctionReflection->getFileName();
 
-        $namespaceNode = null;
-        if (isset($closureData['location']['namespace'])) {
-            $namespaceNode = new Namespace_(new Name($closureData['location']['namespace']));
+        if (\strpos($fileName, 'eval()\'d code') !== false) {
+            throw EvaledClosureCannotBeLocated::create();
         }
 
-        return $this->conversionStrategy->__invoke(
+        FileChecker::checkFile($fileName);
+
+        $fileName = FileHelper::normalizeWindowsPath($fileName);
+
+        $nodeVisitor = new class($fileName, $this->coreFunctionReflection->getStartLine()) extends NodeVisitorAbstract
+        {
+            /**
+             * @var string
+             */
+            private $fileName;
+
+            /**
+             * @var int
+             */
+            private $startLine;
+
+            /**
+             * @var Node\Expr\Closure[]
+             */
+            private $closureNodesData = [];
+
+            /**
+             * @var Namespace_|null
+             */
+            private $currentNamespace;
+
+            public function __construct(string $fileName, int $startLine)
+            {
+                $this->fileName = $fileName;
+                $this->startLine = $startLine;
+            }
+
+            public function enterNode(Node $node) : void
+            {
+                if ($node instanceof Namespace_) {
+                    $this->currentNamespace = $node;
+
+                    return;
+                }
+
+                if ($node instanceof Node\Expr\Closure) {
+                    $this->closureNodesData[] = [$node, $this->currentNamespace];
+                }
+            }
+
+            public function leaveNode(Node $node) : void
+            {
+                if ($node instanceof Namespace_) {
+                    $this->currentNamespace = null;
+                }
+            }
+
+            /**
+             * @return Node[]|null[]|null
+             */
+            public function getClosureNodeData() : ?array
+            {
+                /** @var Node\Expr\Closure[] $closureNodesDataOnSameLine */
+                $closureNodesDataOnSameLine = \array_values(\array_filter($this->closureNodesData, function (array $nodes) : bool {
+                    return $nodes[0]->getLine() === $this->startLine;
+                }));
+
+                if (! $closureNodesDataOnSameLine) {
+                    return null;
+                }
+
+                if (isset($closureNodesDataOnSameLine[1])) {
+                    throw TwoClosuresOnSameLine::create($this->fileName, $this->startLine);
+                }
+
+                return $closureNodesDataOnSameLine[0];
+            }
+        };
+
+        $fileContents = \file_get_contents($fileName);
+        $ast          = (new ParserFactory())->create(ParserFactory::PREFER_PHP7)->parse($fileContents);
```

Comment:
Move this call to the constructor, as I may need to move the parser to an injected dependency shortly